### PR TITLE
 Test ini file common operation  and Update parsecfg.nim 

### PR DESCRIPTION
**Pull request by lihf8515**

### CODE REVIEW

File: lib/pure/parsecfg.nim [hunk 1]
```diff
@@ -359,3 +415,132 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+          c = dict[curSection]
+        c["--" & e.key] = e.value
+        dict[curSection] = c
+      of cfgError:
+        break
+    close(p)
+  result = dict
+
+proc hasNotSymchars(s: string): bool =
+  result = false
```
Comment (Araq) [inline lib/pure/parsecfg.nim]: Use  strutils.allCharsInSet instead.

File: lib/pure/parsecfg.nim [hunk 2]
```diff
@@ -359,3 +415,132 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+        if hasNotSymchars(key[2..key.len()-1]):
+          kv = "--\"" & key[2..key.len()-1] & "\""
+        else:
+          kv = key
+      else:
+        segmentChar = "="
+        kv = key
+      if value != "": ## If the key is not empty
+        if hasNotSymchars(value):
+          kv = kv & segmentChar & "\"" & replace(value) & "\""
```
Comment (Araq) [inline lib/pure/parsecfg.nim]: use kv.add instead

File: lib/pure/parsecfg.nim [hunk 3]
```diff
@@ -359,3 +415,132 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+        else:
+          kv = kv & segmentChar & value
```
Comment (Araq) [inline lib/pure/parsecfg.nim]: use kv.add instead

File: lib/pure/parsecfg.nim [hunk 4]
```diff
@@ -359,3 +415,132 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+      file.writeLine(kv)
+  file.close()
```
Comment (Araq) [inline lib/pure/parsecfg.nim]: use a try finally for the 'close' or a defer.

File: lib/pure/parsecfg.nim [hunk 5]
```diff
@@ -359,3 +415,132 @@ proc next*(c: var CfgParser): CfgEvent {.rtl, extern: "npc$1".} =
...
+
+proc getSectionKey*(dict: OrderedTable[string, OrderedTable[string, string]], section, key: string): string =
+  ## Gets the Key value of the specified Section for a given table.
+  ## Before using it to call the loadCfgFile function.
+  result = ""
+  if dict.haskey(section):
+    if dict[section].hasKey(key):
+      result = dict[section][key]
```
Comment (Araq) [inline lib/pure/parsecfg.nim]: Move the 'result = ""' down into the else (or use a 'return') to save the string allocation in case it has the key.

**Comment by Araq** [comment]
In general OrderedTableRef will give careless programmers (and programmers who value convenience) better performance results and so I think you should use  OrderedTableRef everywhere rather than OrderedTable.

**Comment by lihf8515** [comment]
OK, I'm revising.

**Pull request closed without merging**